spring-vault-core/src/main/java/org/springframework/vault/authentication/LoginTokenAdapter.java

```diff
@@ -104,7 +104,7 @@ static LoginToken augmentWithSelfLookup(RestOperations restOperations,
 		catch (HttpStatusCodeException e) {
 			throw new VaultTokenLookupException(String.format(
 					"Token self-lookup failed: %s %s", e.getStatusCode(),
-					VaultResponses.getError(e.getResponseBodyAsString())));
+					VaultResponses.getError(e.getResponseBodyAsString())), e.getStatusCode());
```

Comment:
This change assumes that lookup is caused by a HTTP status code. Looking at the code above, `VaultTokenLookupException` applies, too with the slight difference that HTTP status was 200 or something other from the 2xx series. We should rather remove the HTTP assumption from `VaultTokenLookupException` and retain the original cause.

Reply:
I think what yiou are saying:
Embed the original HttpStatusCodeException in and remove the HttpCode assumption? Is this a general statement? Sorry I'm not following. I'm ok with embedding the original one, but how is the client to interpret it. I would say with a very rich strata of exceptions (VaultTokenLookup, VaultLoginFailed, etc). But originally in the Issues, you seemed to want to start small.

Reply:
True, the intention was to start small and not to change every single place we're throwing exceptions. I'm fine with the fact that it started growing and we can identify areas where more detailed exceptions make sense.

> Embed the original HttpStatusCodeException in and remove the HttpCode assumption? Is this a general statement?

Having a pull request is great as we see now how creating exception looks and feels and how the changes affect the code. We have exceptions that are purely HTTP related and exceptions that are caused by HTTP interaction but not necessarily bound to HTTP. That's why I mentioned as one of the last comments to use the cause for `VaultHttpException` to transport the actual cause so clients can inspect that if they are interested in that. We shouldn't duplicate functionality for the sake of improved exception cause iteration. 
One thing that isn't covered yet is timeouts and generic I/O errors which are subtypes of `RestClientException`, i.e. `ResourceAccessException`. They don't have a HTTP status yet are they HTTP-related.

If your goal is to inspect each and every HTTP status, then a `RestTemplate` interceptor seems to be the better fit.

> I would say with a very rich strata of exceptions

Very is relative to the observer. Introducing `VaultHttpException` and `VaultLoginException` seems reasonable small to me.